### ISSUE TITLE
Nested block inside if-else fails to parse with "unexpected else"

### ISSUE THREAD

**Issue by francisdb** [unisonweb/unison#324]
```
foo a b =
  if a == "" then
    case Text.length b of
      1 -> false
      _ -> true
  else
    true
```

```
master> 
Token {payload = Open "else", start = Pos 37 3, end = Pos 37 7} :| []
/unison-src/json.u:37:3:
unexpected else
   37 |   else

```

this however works

```
foo a b =
  if a == "" then
    false
  else
    true
```

**Comment by noahhaasis** [comment]
I'm looking into it. The problem is that the lexer doesn't close the case/of block after `true`. The layout stack seems correct so there is a problem closing the blocks.

**Comment by pchiusano** [comment]
Fixed this one. Issue was that for tokens that both start and stop a layout block (like `else` terminates the `then` block and starts a new block), the lexer was always popping just one element from the stack. The correct behavior is to pop until you reach an `Open` token of the correct kind for the closing token (so `else` pops until hitting a `then` token).

Added a test for it as well.

I also spotted another issue that would result in incorrect start/stop annotations for tokens and fixed that in ef3e4d3 too.

Thanks for the report!
